Object bounding box relative to given reference

EliasHasle opened the pull request:
Added optional arguments to setFromObject and expandByObject that allow generating a bounding box relative to a given reference object's coordinate system. Also added some options regarding invisible objects and materials. I consider this an early sketch of the functionality. It should probably be extended to other methods as well, and the logic for undefined materials is probably not perfect.
Related issue: #8222 (visibility)

EliasHasle committed:
Object bounding box relative to given reference

Added optional arguments to setFromObject and expandByObject that allow generating a bounding box relative to a given reference object's coordinate system. Also added some options regarding invisible objects and materials. I consider this an early sketch of the functionality. It should probably be extended to other methods as well, and the logic for undefined materials is probably not perfect.
Related issue: #8222 (visibility)

EliasHasle commented:
I will make a few changes soon. First fix what I think are bugs.

Maybe remove the handling of material existence and visibility, and just stick to the behaviour of traverseVisible.

Maybe modify expandByPoints with a matrix parameter that transforms every vertex before expanding with it, or even add such a matrix parameter to expandByPoint.

Then maybe add an Object3D method that generates the object's own bounding box, optionally in its own coordinates or in world coordinates.

My particular use case is for automatic scaling and alignment of sea vessel models with a standard vessel-oriented coordinate system that can usually be derived from the object-relative bounding box, but not from the world-relative bounding box.

I can solve it using an injection, but I thought this would be useful to others too. Note that the same result cannot be achieved by applying the matrix to the world-relative bounding box.

EliasHasle committed:
Cleanup

Removed "THREE." from instantiations, and added import of Matrix4 in header.
Removed unnecessary conditional function declaration (trav), instead making the choice by manipulating a string. (This would probably matter for some use cases.)
Removed clumsy handling of material visibility. Now it just resorts to the behaviour of traverseVisible.

I have not yet done anything in regards to passing an arbitrary matrix to transform every vertex by before expanding the box with it.

Everything is backwards-compatible.

EliasHasle committed:
Corrected comment

mrdoob commented:
Do you mind elaborating on your use case?

EliasHasle commented:
Not at all. :-)

I let the user load his/her own models, currently in dae or stl format.
There is a standard onboard coordinate system for sea vessels, but the
coordinate system of the loaded file is not guaranteed to correspond to
that. In addition, I want to allow the user to scale the length, breadth
and height of the models using dat.gui sliders.

When the user loads a model, and also when he clicks "autoAlign" after
correcting the model rotation manually, the _autoAlign_ function calculates
the bounding box of the vessel, and positions the model correctly in the
onboard coordinate system.

The dimensions of the bounding box are used as approximations for length,
breadth and height also. My implementation of user-controlled scaling uses
two containers around the model. The internal scaling container _ISCont_
normalizes the model to 1x1x1, and is only modified during _autoAlign_. The
external scaling container is initialized to a scale equal to the
dimensions of the bounding box

31. mars 2017 01:24 skrev "Mr.doob" <notifications@github.com>:

> Do you mind elaborating on your use case?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mrdoob/three.js/pull/11066#issuecomment-290572985>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQaT-wTvSgVTU714zVGG5pA2YVrCtpDoks5rrDmrgaJpZM4MqRov>
> .
>

EliasHasle commented:
, and then the user gets direct control over the scale of _ESCont_ via dat.gui.

I obviously can't use a world bounding box for this. Actually, neither can I use the model's own coordinate system, because it can be rotated, either by definition (at least I have seen ColladaLoader2 set rotation.x to -Math.PI/2) or because the user must correct the rotation before running _autoAlign_. I need to use the coordinate system of a container, and currently use ESCont as reference.

Actually I think my code can be made almost as efficient as the existing one, if one conditionally uses a product of the world matrix and the reference matrix in the beginning of the traversing function, instead of having to do the same check for every vertex and apply both matrices in sequence. I don't think I can make that change until after the weekend, but I think it is promising.

EliasHasle commented:
I think, for generality and efficiency, that maybe the reference parameter should just be the transform matrix. Then one can have a separate method for using an object as reference, that computes the inverse matrix before calling the other method.

EliasHasle commented:
Or, if it is OK code style, one can allow _reference_ to be either an Object3D or a matrix. What do you think?

EliasHasle commented:
One may sometimes want to supply a cqmera ptojection matrix, to get the object's bounding box in screen space, effectively a Box2. E.g., to check qhether an object is contained within a box the user draws on the screen (in RTS games etc.)

EliasHasle commented:
Disclaimer regarding my previous comment: now I am not confident that's exactly how a camera projection matrix works. ;-)

WestLangley commented:
@EliasHasle You can achieve what you want with the existing code by using the following pattern, which adds your object as a child of a parent object.

```
// parent
parent = new THREE.Group();
scene.add( parent );
parent.matrixAutoUpdate = false;

// parent
parent.add( myObject );
```
In the render loop
```
// bounding box in frame of reference of camera
transform = camera.matrixWorldInverse;
transformInverse = camera.matrixWorld;

parent.matrix.copy( transform );
boxHelper.update( parent );
parent.matrix.identity();

// convert box to world coordinates for rendering
transformInverse.applyToBufferAttribute( boxHelper.geometry.attributes.position );
```
Use an orthographic camera for rendering.

EliasHasle commented:
@WestLangley
Thanks for the tip. At first glance , the solution looks more complicated
and less flexible than my approach, but I will investigate it a bit when I
get back to this.

I know this PR could be completed and perfected in less than an hour, but I
have just started a mental "detox" because I have been spending too much of
my time and resources thinking about 3D programming lately. So... I'll be
back. :-)

EliasHasle committed:
Separated out method using matrix, and optimized.

Now the user can choose to supply any Matrix4 to the expandByTransformedObject method, to set the coordinate system that the bounding box will be computed in.

EliasHasle commented:
So @mrdoob , if you like it, I can update the docs too, and possibly make some corresponding additions to the (bounding) box helper(s).

The testing I have done is very limited. Will I also have to create unit tests?

(The recent change of allowing the user to pass an arbitrary matrix that is applied to all points (in world space) almost tempts me to make a similar method that allows the user to supply any function from Vector3 to Vector3... I mean, I can imagine use cases, but then again I can do with an injection too.)

EliasHasle committed:
Oops, had to remove "THREE." again

mrdoob commented:
I think I prefer @WestLangley approach, which doesn't require library changes.
